### PR TITLE
Task level retry and Stage level retry

### LINKED ISSUE #140: Failed task retry
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

When a task is failed on Ballista Executor, due to different failure reasons, sometimes the failed tasks need to be resubmitted to another Executor and rerun the failed tasks. Need to enhance the Ballista Scheduler to support this.
Need to enhance `FailedTask` rpc message to indicate whether the failure task should retry or not.
A good example of failed task that should retry is Executor lost. And on the other hand, user cancelled queries/tasks should not retry. A shuffle read failure should not retry the task either.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by mingmwang**
# Which issue does this PR close?

Closes #140.

 # Rationale for this change

# What changes are included in this PR?

**1. Task level retry,** 
a) Add task attempt num, use a job level unique task_id to represent the Task
b) Define couple of failure reasons for FailedTask

````
pub enum FailedReason {
        #[prost(message, tag="4")]
        ExecutionError(super::ExecutionError),
        #[prost(message, tag="5")]
        FetchPartitionError(super::FetchPartitionError),
        #[prost(message, tag="6")]
        IoError(super::IoError),
        #[prost(message, tag="7")]
        ExecutorLost(super::ExecutorLost),
        /// A successful task's result is lost due to executor lost
        #[prost(message, tag="8")]
        ResultLost(super::ResultLost),
        #[prost(message, tag="9")]
        TaskKilled(super::TaskKilled),
    }

````
c) Reasoning the real task failure reason when return the TaskStatus back to Scheduler
d) Based on the failure reason, scheduler decide to reschedule the task and bump task attempt number of failed task.

**2. Stage level retry and shuffle read failure handling** 
a) Add stage attempt num.
b) When there are shuffle partition fetch failures, the current running stage will be rolled back and the map stages will be resubmit.
c) handling of delayed fetch failure task updates
d) Cancel the running tasks if Scheduler decide to fail the stage/job.

And below two items are not covered in this PR and we may need to revisit the related logic in future:
1) If the plans/expressions in the Stage is not deterministic, need to revisit the resubmit logic.
2) If we have a map stage whose shuffle output can be reused by multiple reduce stages, need to revisit the stage retry logic.

# Are there any user-facing changes?

### CODE REVIEW

**Comment by mingmwang** [comment]
@thinkharderdev @andygrove @yahoNanJing
Please help to review my PR.

**Comment by mingmwang** [comment]
Added 10+ UTs to cover different cases.

**Comment by yahoNanJing** [comment]
With this PR, the state machine for the stage becomes as follows:
StageStateMachine

The transitions marked in red are newly added ones to deal with error task recovery.

**Comment by yahoNanJing** [comment]
For example, one SQL stage graph snapshot is as follows:

           							Stage 4(Resolved)
                                                  ↗                                                                       ↘
Stage 1(Successful)                										Stage 5(Unresolved)
                                       ↘                                                                                   ↗
           				      Stage 2(Successful)  ->  Stage 3(Running) 

Both of Stage1 and Stage2 has output shuffle data resides on Executor1. Unluckily, the Executor1 gets lost. Then tasks for the running Stage3 will fail due to not being able to fetch shuffle data. For this kind of scenario and error, by this PR, the ExecutionGraph will be able to continue running with proper task reset.

File: ballista/rust/scheduler/src/state/execution_graph.rs
Comment (yahoNanJing) [inline ballista/rust/scheduler/src/state/execution_graph.rs:471]: Here, only failed tasks will be dealt with. Once a running stage converted into an unresolved, it will discard all of its tasks info. Then to update the successful task info is meaningless.

**Comment by thinkharderdev** [comment]
@mingmwang I'm pretty swamped this week so won't have time to review this until this weekend.

File: ballista/rust/scheduler/src/state/execution_graph.rs
Comment (yahoNanJing) [inline ballista/rust/scheduler/src/state/execution_graph.rs:429]: When this running stage is converted from successful stage, there may be in flight running tasks of the stages depending on this stage. When we find those tasks fails due to fetching data failure, then we also need to reset this running stage. Therefore, here, the reset_running_stages check is necessary.

**Comment by yahoNanJing** [comment]
Later, maybe we can make this error task recovery feature configurable.

**Comment by yahoNanJing** [comment]
The overall design of this PR is good for me and it will be really useful to deal with executor lost issue and executor bad disk issue. @thinkharderdev, @andygrove, @avantgardnerio, could you help review this PR?

**Comment by andygrove** [comment]
Thanks, @mingmwang. I plan on starting to review this tomorrow.

File: ballista/rust/core/src/execution_plans/shuffle_reader.rs
Comment (yahoNanJing) [inline ballista/rust/core/src/execution_plans/shuffle_reader.rs:266]: Trait is much more extensible than condition compiling 👍

File: ballista/rust/core/proto/ballista.proto
Comment (andygrove) [inline ballista/rust/core/proto/ballista.proto:427]: nit: `tid_gen` isn't a descriptive name. I am guessing this is short for `task_id_gen`?

File: ballista/rust/core/proto/ballista.proto
Comment (andygrove) [inline ballista/rust/core/proto/ballista.proto:559]: Could you add comments explaining these different partition ids?

**Comment by andygrove** [comment]
@mingmwang I have taken a first pass through this PR and I think it looks good. I am going to spend some time testing out the PR locally. It would be good to wait for Dan to review as well before we merge this.

File: ballista/rust/core/proto/ballista.proto
Comment (andygrove) [inline ballista/rust/core/proto/ballista.proto:458]: There are breaking changes here because this PR is changing the meaning of fields. For example field 3 was `PhysicalHashRepartition` and now it is `uint32`. We should assign new numbers to new fields and not modify numbers for existing fields.

File: ballista/rust/core/proto/ballista.proto
Comment (andygrove) [inline ballista/rust/core/proto/ballista.proto:481]: Same issue here with changing tag numbers.

File: ballista/rust/core/proto/ballista.proto
Comment (andygrove) [reply ballista/rust/core/proto/ballista.proto:458]: These changes meant that I could not use my benchmark client without recompiling it against this branch, which should not be necessary.

**Comment by andygrove** [comment]
I tested this locally, and it worked really well!

I ran one scheduler, and two executors and I could kill one executor and still see a query complete successfully. This is not the case in the master branch.

I am happy to approve this once feedback has been addressed.

File: ballista/rust/core/proto/ballista.proto
Comment (mingmwang) [reply ballista/rust/core/proto/ballista.proto:427]: Sure,  will change to `task_id_gen`

File: ballista/rust/core/proto/ballista.proto
Comment (mingmwang) [reply ballista/rust/core/proto/ballista.proto:458]: > These changes meant that I could not use my benchmark client without recompiling it against this branch, which should not be necessary.

Well, I thought those are private structures and only used internally by the TaskManager.

**Comment by mingmwang** [comment]
> I tested this locally, and it worked really well!
> 
> I ran one scheduler, and two executors and I could kill one executor and still see a query complete successfully. This is not the case in the master branch.
> 
> I am happy to approve this once feedback has been addressed.

Thank you. We will also start the chaos-monkey testing in next month to verify the recent changes.

File: ballista/rust/core/proto/ballista.proto
Comment (mingmwang) [reply ballista/rust/core/proto/ballista.proto:559]: Sure, will add more comments in the code. Here the new added `map_partition_id` is the partition_id of the map stage who produce this shuffle data.  But the original `PartitionId partition_id` has different meanings in different places.
Sometimes it stands for a task_id, and here it stands for a shuffle partition id(composition of map_stage_id + reduce partition id + job_id), a mixed up of map and reduce infos together.
So if we do not consider the backward compatibility, I would suggest to unnest this and make the PartitionLocation a plan struct like blow:

````
message PartitionLocation {
  string job_id = 1;
  uint32 map_stage_id = 2;
  uint32 map_partition_id = 3;
  uint32 partition_id = 4;
  ExecutorMetadata executor_meta = 5;
  PartitionStats partition_stats = 6;
  string path = 7;
}
````

File: ballista/rust/core/proto/ballista.proto
Comment (mingmwang) [reply ballista/rust/core/proto/ballista.proto:458]: I will change the tag num back.  But still there are backward incompatible changes here,  The `stage_id` was changed from `uint64 stage_id` to `uint32 stage_id` to have the same type with other stage_ids in this proto file.

Comment (andygrove) [reply ballista/rust/core/proto/ballista.proto]: Yes, after reviewing again, I don't think we can avoid breaking changes given the extent of these changes.

File: ballista/rust/core/proto/ballista.proto
Comment (andygrove) [reply ballista/rust/core/proto/ballista.proto]: It is late here for me, but I think that makes sense. I will review it again tomorrow.

**Review by andygrove — APPROVED** [review]
Thanks @mingmwang. This is a very nice improvement in stability.

**Review by thinkharderdev — APPROVED** [review]
Had a few questions but this looks really great! Thanks for your work on this @mingmwang!

File: ballista/rust/core/src/error.rs
Comment (mingmwang) [reply ballista/rust/core/src/error.rs:263]: Yes, have both `retyrable` `and count_to_failures` is to support the case that we might have some specific error that we want it to retry forever until it is successful.

File: ballista/rust/core/src/execution_plans/shuffle_reader.rs
Comment (mingmwang) [reply ballista/rust/core/src/execution_plans/shuffle_reader.rs:313]: > Nice! I assume where this is going is making the executors smart enough to read partitions on the same physical machine directly from disk.

Yes, exactly.

File: ballista/rust/scheduler/src/state/execution_graph.rs
Comment (mingmwang) [reply ballista/rust/scheduler/src/state/execution_graph.rs:121]: > I don't get why we are saving a `HashSet` of attempts. Shouldn't it just be `usize`?

The purpose of using HashSet is to record what are the exact distinct failed attempts.

File: ballista/rust/core/proto/ballista.proto
Comment (mingmwang) [reply ballista/rust/core/proto/ballista.proto:738]: > Not sure I understand the difference between these two errors

Yes, good question. In the current code base, both the two errors are not used directly by the executor tasks.
They are used by the Scheduler. When we see a 'FetchPartitionError' task update from the reduce task, the related map task's status is changed to  'ResultLost'.  Of cause most of the time `ResultLost` should be caused by `ExecutorLost`.

**Comment by yahoNanJing** [comment]
Thanks @mingmwang for this huge step of error recovering. Thanks @andygrove and @thinkharderdev for reviewing this PR. Since we all approved this PR, I'll merge it first.

**Review by yahoNanJing — APPROVED** [review]

**Pull request merged**
